core/client/fs/src/main/java/alluxio/client/file/FileSystemUtils.java

```diff
@@ -122,53 +119,43 @@ public static boolean waitCompleted(final FileSystem fs, final AlluxioURI uri,
         timeleft = deadline - System.currentTimeMillis();
       }
     }
-
     return completed;
   }
 
   /**
-   * Persists the given file to the under file system.
+   * Convenience method for {@code #waitPersisted(fs, uri, -1)}. i.e. wait for an indefinite period
+   * of time to persist. This will block for an indefinite period of time if the path is never
+   * persisted. Use with care.
    *
    * @param fs {@link FileSystem} to carry out Alluxio operations
-   * @param fsContext the {@link FileSystemContext} linked to the {@link FileSystem} client
    * @param uri the uri of the file to persist
    */
-  public static void persistFile(final FileSystem fs, final FileSystemContext fsContext,
-      final AlluxioURI uri)
+  public static void waitPersisted(final FileSystem fs, final AlluxioURI uri)
       throws IOException, TimeoutException, InterruptedException {
-    FileSystemMasterClient client = fsContext.acquireMasterClient();
-    try {
-      client.scheduleAsyncPersist(uri);
-    } finally {
-      fsContext.releaseMasterClient(client);
-    }
+    waitPersisted(fs, uri, -1);
+  }
+
+  /**
+   * Persists the given path to the under file system and returns once the persist is complete.
+   * Note that if this method times out, the persist may still occur after the timeout period.
+   *
+   * @param fs {@link FileSystem} to carry out Alluxio operations
+   * @param uri the uri of the file to persist
+   * @param timeoutMs max amount of time to wait for persist in milliseconds. -1 to wait
+   *                  indefinitely
+   * @throws TimeoutException if the persist takes longer than the timeout
+   */
+  public static void waitPersisted(final FileSystem fs, final AlluxioURI uri, int timeoutMs)
```

Comment:
`waitPersisted` sounds like it would wait for something to be persisted, as opposed to persist something and then wait. Perhaps we could name it `persistAndWait`? Alternately, we could have callers call `persist` themselves before calling `waitPersisted`. This has the advantage of letting us specify separate options for persistence vs waiting for persistence

Reply:
I agree that we should probably change the names but I'm not quite sure what I want to rename them to. Another option is calling the client API method `asyncPersist` and the then we could put something like `synchronousPersist` in `FileSystemUtils`?

Calling the user-facing method `asyncPersist` removes any ambiguities on how the method behaves, but removes the flexibility of changing the behavior of the method in the future.

I'm not a huge fan of `persistAndWait` as a method name. It just sounds clunky. I also don't like `waitPersisted`

Reply:
So I see 3 options (ordered by my preference):

1. API method gets `asyncPersist`, `FileSystemUtils` gets `synchronousPersist`
2. API method remains `persist`, `FileSystemUtils` drops the persist action and we call it `waitPersisted`. It polls for a file's persistence state as opposed to persisting, and then polling
3. API method remains `persist`, `FileSystemUtils` gets renamed to `persistAndWait`